### DISCUSSION TITLE
Views to restrict access to columns?

### DISCUSSION THREAD

**Discussion by boyeln** [supabase/supabase#1275] [Questions] [answered]
Hi, and thanks for an incredible platform!

In [the blog post on postgres views](https://supabase.io/blog/2020/11/18/postgresql-views) under "Why should we use Views?" -> "Security" it says:

> Views can restrict the amount and type of data presented to a user. Instead of allowing a user direct access to a set of tables, we provide them a view instead. We can prevent them from reading sensitive columns by excluding them from the underlying query.

How exactly can that be achieved? Let's say for a simple polling app, you have polls and options tables like this (there's probably syntax errors, but you get the gist of it):

```sql
create table polls
  id bigint generated by default as identity primary key,
  question text not null,
  vote_slug varchar(16) unique not null,
  result_slug varchar(16) unique not null,
  user_id uuid references auth.user(id);

create table options
  id bigint generated by default as identity primary key,
  poll_id bigint not null references polls(id),
  votes bigint not null default 0;
```

Where `vote_slug` is the slug for the sharable link where people can vote, and `result_slug` is the slug for the sharable link where the results are visible. You want the polls table to be publicly accessible, but you don't want everyone to see the `result_slug`. Can this be achieved with a view? As far as i know, row level security is only available for the table, and not for views (I assume they'll use the table's row level security).

One solution is of course to extract the slugs to their own table and restrict it that way, but it would be really nice if I could have one table less. Or does it exist some other (better?) technique for restricting access to columns?

**Comment by silentworks** [comment, marked as answer]
Yes this can be achieved with Views, remember that views are just predefined queries that create a sort of virtual table inside of Postgres. Your view code would be written as:

```sql
CREATE VIEW poll_without_results AS
  SELECT 
    id, 
    question, 
    vote_slug, 
    user_id
  FROM polls
```
Then inside of your supabase js code:

```js
await supabase.from('poll_without_results')
  .select('*');
```

This would return a view that doesn't contain any data related to the result_slug at all. Remember above I said Views create a virtual table in Postgres, hence why you can use the same `.from` and `.select` as you would with the supabase js api.

**Reply by boyeln** [reply]
I probably didn't clarify this well enough in the question, but the expected/intended result is that a user can query all the polls, but not be able to see the "sensitive" field `result_slug`. By just creating a view, the field can still be accessed by users by just querying the `polls` table rather than the `poll_without_results`. And if I enable `row level security` on the `polls` table and restrict access using a policy to only the polls a user owns, this would also be applied to the view. Ideally I would enable `row level security` on the view itself, and have a different policy on the view. But as you state, views are just predefined queries, thus not supporting `row level security`, only applying the existing security on the original table(s).

So I guess my question is more how views can be used for security in the way described in the blog. I understand that they can be used for convenience (and performance gains), but not how they can be used for added security.

**Reply by silentworks** [reply]
No restricting access on the polls table to only polls users own would not affect the view, remember the view is its own table. My suggestion above fits what you are after, you just need to add the row level security and its all good. I have tested this in a playground repository I have and it works as how you described it above.

A non-logged in user can view all polls without the results and a logged in user can view polls including the results slug. You could also restrict this from them too by just restricting access to the `polls` table completely and use predefined views to show a set of results per logged in user type.

I have attached a video below showing how this is working at the moment, I will push this playground repository online soon and you can dive into the code yourself, the repository contains a bunch of migration scripts that will show what policy is in use too.

https://user-images.githubusercontent.com/79497/116108186-342a4e80-a6a3-11eb-83ef-a430ad6a1f4d.mp4

**Reply by boyeln** [reply]
Thanks so much for such a comprehensive answer! This is really awesome!

**Reply by silentworks** [reply]
I'm not sure what you are asking, its also best to create a new topic on the discussions related to what you are asking than adding to one that has already been answered.

**Reply by savikko** [reply]
> I will push this playground repository online soon 

@silentworks : I tried to find this repository, but could not. I would love to see how that policy is build.

**Reply by juanvilladev** [reply]
Thanks for providing this example @silentworks 

In my opinion, developers new to Postgres should be careful when using views. As you might assume that views respect the underlying RLS by default, but that is not the case. Views are by default security definer, thus if you define them using the Supabase Admin role from the dashboard. Any user can access that data. 

If within the case of @boyeborg users needed to be authenticated before view ANY of the columns, this resolution would have caused a security issue as anon users could query the view. 

@boyeborg the assumption you made "As far as i know, row level security is only available for the table, and not for views (I assume they'll use the table's row level security)." is not true unless you define the view with security_invoker = true

With Postgres 15, one can define a view with security_invoker = true:
```
CREATE VIEW WITH (security_invoker = true) AS 
   SELECT....
```
This will apply the RLS of the underlying tables to the view.

@silentworks example works BECAUSE they view is by default as SECURITY_DEFINER. 
He can enable RLS on it, however the view is bypassing it. This example is only valid if you want to make your view visible to anyone including anon.

**Reply by pbell23** [reply]
@juanvilladev @silentworks  Do you know how we can make the view only visible to authenticated role and not anon while still preventing viewing the underlying table ?

**Reply by zubaird** [reply]
Thanks and from what I can tell, in a situation where you have some columns useful for public view, but others that must be restricted or even totally private, it is desirable that Views use the Supabase admin role when presenting the data. This allows you to keep your RLS even more restrictive while making the public ok columns you want to show available. 

Oh @pbell23  in your case I think you'll need to include the security_invoker = true, which will use your RLS that you setup for authenticated access.
